### ISSUE TITLE
Unexpected behaviour of sparsemax gradients for 3d tensors 

### ISSUE THREAD

**Issue by mtreviso** [deep-spin/entmax#13]
Hi folks! 

It seems like the gradients of sparsemax are not the same when we have two "equal" tensors: one 2d, and the other with a time dimension.

Here is the code to reproduce the problem:
```python
import torch
import entmax


def test_map_fn(activation_fn):
    x = torch.tensor([[-2, 0, 0.5], [0.1, 2, -0.4]], requires_grad=True)
    # >>> x.shape
    # torch.Size([2, 3])
    a_2d = activation_fn(x, dim=-1)
    z_2d = torch.sum(torch.pow(a_2d, 2))
    z_2d.backward()
    grad_2d = x.grad

    x = torch.tensor([[[-2, 0, 0.5]], [[0.1, 2, -0.4]]], requires_grad=True)
    # >>> x.shape
    # torch.Size([2, 1, 3])
    a_3d = activation_fn(x, dim=-1)
    z_3d = torch.sum(torch.pow(a_3d, 2))
    z_3d.backward()
    grad_3d = x.grad

    print(activation_fn.__name__)
    print('Ok acts:', torch.allclose(a_2d.squeeze(), a_3d.squeeze()))
    print('Ok grads:', torch.allclose(grad_2d.squeeze(), grad_3d.squeeze()))
    print(grad_2d.squeeze())
    print(grad_3d.squeeze())
    print('---\n')


if __name__ == '__main__':
    test_map_fn(torch.softmax)
    test_map_fn(entmax.entmax15)
    test_map_fn(entmax.sparsemax)
```

The output of this code is:
```
softmax
Ok acts: True
Ok grads: True
tensor([[-0.0421, -0.0883,  0.1304],
        [-0.1325,  0.2198, -0.0873]])
tensor([[-0.0421, -0.0883,  0.1304],
        [-0.1325,  0.2198, -0.0873]])
---

entmax15
Ok acts: True
Ok grads: True
tensor([[ 0.0000, -0.2344,  0.2344],
        [-0.0926,  0.0926,  0.0000]])
tensor([[ 0.0000, -0.2344,  0.2344],
        [-0.0926,  0.0926,  0.0000]])
---

sparsemax
Ok acts: True
Ok grads: False
tensor([[ 0.0000, -0.5000,  0.5000],
        [ 0.0000,  0.0000,  0.0000]])
tensor([[ 0., -2.,  0.],
        [ 0.,  1.,  0.]])
---
```

So, using sparsemax, the grads of the two tensors are different. Obs: it seems that a quick fix by doing `tensor.view(-1, nb_labels)` to get a 2d tensor works fine in practice.

**Comment by vene** [comment]
Hi @mtreviso ,

Interesting catch, does it happen also with entmax_bisect with alpha=2?

I think we have unit tests that should cover the correctness of gradients for multi-dimensional inputs. If you have time and willingness, could you check which test should be catching this and why it isn't?

**Comment by mtreviso** [comment]
With entmax_bisect with alpha=2 it doesn't happen! I tested with sparsemax_bisect and it is ok too.

**Comment by mtreviso** [comment]
Turns out if you create a tensor of shape (2, 2, 3) the grad is actually correct. I guess the problem is in [this line](https://github.com/deep-spin/entmax/blob/master/entmax/activations.py#L163). We should squeeze the dimension that we are working on. 

I tested here with 
`v_hat = grad_input.sum(dim=dim) / supp_size.to(output.dtype).squeeze(dim)` and it seems to be working fine. 

I think the tests in `rest_root_finding.py`, namely `test_sparsemax_grad`, `test_entmax_grad` didn't catch this because they are testing with tensors with two non-unitary dims, and the tests are just for the bisect implementation of both sparsemax and entmax.

**Comment by vene** [comment]
test_root_finding isn't supposed to test this impl but the bisection-based one. (this implementation is not based on root finding)

I think the correct line you want is the one from the entmax15 backward pass:
https://github.com/deep-spin/entmax/blob/master/entmax/activations.py#L189

**Comment by mtreviso** [comment]
Got it. So, should I create a special test file only for non root finding activations and do a pr?

Yeah! entmax backward pass looks ok to me (in terms of tensor shapes). I think is just a matter of calling squeeze with dim=ctx.dim in the sparsemax backward pass, right?

**Comment by vene** [comment]
So since the entmax15 backward pass does not call squeeze and acts correctly, I suspect squeeze is not needed for sparsemax either. This is why I linked you to the entmax15 backward pass.

The tests for the topk-based exact solvers are in https://github.com/deep-spin/entmax/blob/master/entmax/test_topk.py and it seems none are checking the gradient rn.

**Comment by mtreviso** [comment]
I think we actually have to squeeze `supp_size` in the correct dimension, since it is unsqueezed before in this line:
https://github.com/deep-spin/entmax/blob/master/entmax/activations.py#L71

I tried your suggestion and I just removed `.squeeze()` and left `supp_size` as it is. If you create a tensor of shape (3,2,2) and try to apply sparsemax with dim=-1, you'll get RuntimeError.

Using entmax:
```python
>>> x = torch.randn(3,2,2, requires_grad=True)
>>> z = torch.sum(torch.pow(entmax.entmax15(x), 2))
>>> z.backward()
>>> x.grad.shape
torch.Size([3, 2, 2])
```

sparsemax:
```python
>>> x = torch.randn(3,2,2, requires_grad=True)
>>> z = torch.sum(torch.pow(entmax.entmax15(x), 2))
>>> z.backward()
>>> x.grad.shape
  File "/home/mtreviso/Documents/entmax/entmax/activations.py", line 164, in backward
    v_hat = grad_input.sum(dim=dim) / supp_size.to(output.dtype)
RuntimeError: The size of tensor a (3) must match the size of tensor b (2) at non-singleton dimension 1
```

Inspecting `grad_input.sum(dim=dim).shape` and `supp_size.shape` we see that they are not broadcastable:
```python
>>> grad_input.sum(dim=dim).shape
torch.Size([3, 2])
>>> supp_size.shape
torch.Size([3, 2, 1])
```

Maybe writing `test_sparsemax_grad()` and `test_entmax_grad()` in `test_topk.py`?

**Comment by mtreviso** [comment]
Closing. Solved by #18.
